### PR TITLE
Fixes #35585 - add an option to enable the callback plugin for job templates

### PR DESCRIPTION
**Pull request by nofaralfasi**
Added the property ansible_callback_enabled to the templates table. Created a Pagelet on the template's form to add the callback option to the UI.

Requires:
- https://github.com/theforeman/foreman/pull/9503

A few questions regarding these changes:
1. Do we want to allow to change the callback-plugin value if the template is locked?
2. Should we add a check for the `Job Category` type (if it is for Ansible), to decide whether to display the new ansible-callback tab? 
3. Is this change requires updating the plugin's version? 

A screenshot with the new tab in the UI:
image

### CODE REVIEW

File: db/migrate/20221003153000_add_ansible_callback_enabled_to_templates.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+class AddAnsibleCallbackEnabledToTemplates < ActiveRecord::Migration[6.0]
+  def change
+    add_column :templates, :ansible_callback_enabled, :boolean, default: false
+    Template.where(name: 'Ansible Roles - Ansible Default').update_all(ansible_callback_enabled: true)
```
Comment (adamruzicka) [inline db/migrate/20221003153000_add_ansible_callback_enabled_to_templates.rb:6]: In here we should also find the templates which are assigned to `ansible_run_host` feature and set the flag for them as well.

We could even go one step further and set the flag only for the templates with the feature to not introduce any surprising behavior.

File: app/models/foreman_ansible/ansible_provider.rb [hunk 2]
```diff
@@ -122,8 +122,7 @@ def proxy_batch_size
         private
 
         def ansible_command?(template)
-          template.remote_execution_features.
-            where(:label => 'ansible_run_host').empty?
+          !template.ansible_callback_enabled
```
[the remark this answers is not in the document]
Comment (nofaralfasi) [reply app/models/foreman_ansible/ansible_provider.rb:125]: But in the UI we still give the option to check/uncheck the `enable ansible callback` option. 
Wouldn't that be confusing? 

If you say it's OK then I will keep both of them, just making sure.

File: db/migrate/20221003153000_add_ansible_callback_enabled_to_templates.rb (hunk 1 above)
Comment (nofaralfasi) [inline db/migrate/20221003153000_add_ansible_callback_enabled_to_templates.rb:6]: @adamruzicka did you mean doing this:
` Template.find(RemoteExecutionFeature.where(label: 'ansible_run_host').first.job_template_id).update_all(ansible_callback_enabled: true)`?

Comment (adamruzicka) [reply db/migrate/20221003153000_add_ansible_callback_enabled_to_templates.rb:6]: Almost. The way the database schema is laid out, it is possible that one feature will have more than one template assigned to it.

File: app/models/foreman_ansible/ansible_provider.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (adamruzicka) [reply app/models/foreman_ansible/ansible_provider.rb:125]: >  Wouldn't that be confusing?

That's true, but still probably worth it over breaking existing workflows.

File: db/migrate/20221003153000_add_ansible_callback_enabled_to_templates.rb (hunk 1 above)
Comment (adamruzicka) [reply db/migrate/20221003153000_add_ansible_callback_enabled_to_templates.rb:6]: > You mean that we should give the option to enable the callback only for templates assigned to the ansible_run_host feature?

Yes, but only in this migration

> And also implement it in the UI?

No, users should be free to do whatever they please. It is just that the migration should honor whatever changes the users might have already made

**Comment by nofaralfasi** [comment]
Hi @adamruzicka, I updated some of the things we talked about, but not all.
I have some changes related to the question I asked here https://github.com/theforeman/foreman/pull/9503#discussion_r1022588995, and also I couldn't find a way yet to add the new `ansible_callback_enabled` field to the `job_templates` API.

**Review by adamruzicka — CHANGES_REQUESTED** [review]

File: lib/foreman_ansible/register.rb [hunk 3]
```diff
@@ -257,4 +259,16 @@ def ansible_variables_with_overrides
                         :partial => 'foreman/smart_proxies/update_smart_proxy',
                         :onlyif => ->(proxy, view) { view.can_update_proxy?(proxy) }
   end
+  extend_page('templates/_form') do |context|
+    context.add_pagelet :tab_headers,
+                        :name => _('Ansible Callback'),
+                        :partial => 'job_templates/job_template_callback_tab_headers',
+                        :onlyif => ->(subject, view) { view.correct_template? }
```
Comment (nofaralfasi) [inline lib/foreman_ansible/register.rb:266]: @adamruzicka is there a way to implement this method call in one line? (without having to create the `JobTemplatesHelper`?)

Comment (nofaralfasi) [reply lib/foreman_ansible/register.rb]: For some reason I am getting that `subject` is nil

Comment (adamruzicka) [reply lib/foreman_ansible/register.rb]: Ah, interesting. It works for `:tab_content`. Most likely because you are not providing the subject for the headers in here https://github.com/theforeman/foreman/pull/9503/files#diff-652d7608bcc9387e191b8c1a53a8cf45bb75cbf21e776863d7b14ca206d9774fR15

Comment (nofaralfasi) [reply lib/foreman_ansible/register.rb]: I found out why. I didn't pass it in foreman core, when calling the `render_pagelets_for(:tab_headers)`.

**Comment by adamruzicka** [comment]
Personally I'd just ignore it. Rubocop thinks it is good enough.

If you wanted to do it though, this would be one of the possible approaches:
1) Identify a large chunk of `register.rb` - good candidates would be the `settings` block, `security_block` or the `extend_page` calls
2) Move it to a separate file (and module as a `self.something` method)
3) Require the file in `register.rb`
4) Call the module from step 2 inside the `Foreman::Plugin.register` block and pass self to it

To me going through all this isn't really worth it. The changes would be mostly unrelated to the actual changes you're trying to do here and as far as I know codeclimate isn't really used all that much in the wider foreman ecosystem.

**Comment by nofaralfasi** [comment]
> To me going through all this isn't really worth it. The changes would be mostly unrelated to the actual changes you're trying to do here and as far as I know codeclimate isn't really used all that much in the wider foreman ecosystem.

Then I will leave it that way. Thanks for the explanation!

File: app/views/job_templates/_job_template_callback_tab_headers.html.erb [hunk 4]
```diff
@@ -0,0 +1 @@
+<li><a href="#ansible_callback_enabled" data-toggle="tab"><%= _("Ansible Callback") %></a></li>
```
Comment (adamruzicka) [inline app/views/job_templates/_job_template_callback_tab_headers.html.erb:1]: The name predestines the entire tab to never having anything else in it apart from the single checkbox. How about we go with more generic "Ansible"?

File: app/views/job_templates/_job_template_callback_tab_content.html.erb [hunk 5]
```diff
@@ -0,0 +1,3 @@
+<div class="tab-pane" id="ansible_callback_enabled">
+  <%= checkbox_f f, :ansible_callback_enabled, :label=>_('Enable Ansible Callback'), :disabled => @template.locked? %>
```
Comment (adamruzicka) [inline app/views/job_templates/_job_template_callback_tab_content.html.erb:2]: Spaces around the rocket after `:label`

Comment (adamruzicka) [inline app/views/job_templates/_job_template_callback_tab_content.html.erb:2]: It is somewhat unfortunate that right now the tab is displayed even if the provider type is not ansible. I was able to put together a makeshift solution with jquery, although I don't really dare to suggest you to go down that path.

@MariaAga If we have an old erb page that we're extending from a plugin using pagelets and we want the content to dynamically change based on field from the original, non-extended page, do we have any other option apart from registering an onchange handler onto the fields we're interested in?

Comment (MariaAga) [reply app/views/job_templates/_job_template_callback_tab_content.html.erb:2]: pagelets are really not my area, so we might have to use that suggested jquery

File: app/views/job_templates/_job_template_callback_tab_headers.html.erb [hunk 6]
```diff
@@ -0,0 +1,13 @@
+<li><a id="ansible_tab_header" href="#ansible_callback_enabled" data-toggle="tab"><%= _("Ansible") %></a></li>
+<script type="text/javascript">
+    $(document).ready(function () {
+        var provider_type = $('#job_template_provider_type');
+        provider_type.change(ansibleTabVisibilityByProvider);
+        provider_type.change();
+
+        function ansibleTabVisibilityByProvider() {
+            var tab_header = $("#ansible_tab_header");
+            $("#select2-chosen-1").text() === 'Ansible' ?  tab_header.show() : tab_header.hide();
```
Comment (nofaralfasi) [inline app/views/job_templates/_job_template_callback_tab_headers.html.erb:10]: @adamruzicka this is the jquery code I added. 
Do you think we should add this line of code to the false statement (rewrite the if condition):
` $("#job_template_ansible_callback_enabled").prop('checked', false);`

This is for cases the user sets the `ansible_callback_enabled` to true and then changes the provider to `Script`. I'm not sure if it's necessary.

Comment (adamruzicka) [reply app/views/job_templates/_job_template_callback_tab_headers.html.erb:10]: That's probably not necessary. They flag might be set, but if the provider_type is not ansible, it will not be used in any way.

File: app/views/job_templates/_job_template_callback_tab_headers.html.erb [hunk 7]
```diff
@@ -0,0 +1,13 @@
+<li><a id="ansible_tab_header" href="#ansible_callback_enabled" data-toggle="tab"><%= _("Ansible") %></a></li>
+<script type="text/javascript">
+    $(document).ready(function () {
+        var provider_type = $('#job_template_provider_type');
+        provider_type.change(ansibleTabVisibilityByProvider);
+        provider_type.change();
```
Comment (adamruzicka) [inline app/views/job_templates/_job_template_callback_tab_headers.html.erb:6]: what does this do?

File: app/views/job_templates/_job_template_callback_tab_headers.html.erb (hunk 6 above)
Comment (adamruzicka) [inline app/views/job_templates/_job_template_callback_tab_headers.html.erb:10]: This function can take an argument - the element that changed. From it we can pull the value without resorting to touching select2's internals.

File: app/views/job_templates/_job_template_callback_tab_headers.html.erb (hunk 7 above)
Comment (nofaralfasi) [reply app/views/job_templates/_job_template_callback_tab_headers.html.erb:6]: The first call (`provider_type.change(ansibleTabVisibilityByProvider)`) attaches the function to execute when a change event occurs.
The second call (`provider_type.change()`) triggers a change event (to call the `ansibleTabVisibilityByProvider` function when the page loads).

**Comment by adamruzicka** [comment]
I don't think the change in foreman will be picked to 3.5 so maybe bumping the required foreman version would be in order?

Also I'm not sure what the test failures are about

**Comment by nofaralfasi** [comment]
> Also I'm not sure what the test failures are about

I updated foreman dependency, and now the tests are passing.

File: app/views/job_templates/_job_template_callback_tab_content.html.erb [hunk 8]
```diff
@@ -0,0 +1,3 @@
+<div class="tab-pane" id="ansible_callback_enabled">
+  <%= checkbox_f f, :ansible_callback_enabled, :label => _('Enable Ansible Callback'), :disabled => @template.locked? %>
```
Comment (nofaralfasi) [inline app/views/job_templates/_job_template_callback_tab_content.html.erb:2]: @adamruzicka do we need to add translations for that? (I've never done this so I'm not sure what it includes).

File: app/controllers/foreman_ansible/api/v2/job_templates_controller_extensions.rb [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+module ForemanAnsible
+  module Api
+    module V2
+      # Extends the job_templates api controller to support creating/updating with ansible callback plugin
+      module JobTemplatesControllerExtensions
+        extend Apipie::DSL::Concern
+
+        update_api(:create, :update) do
+          param :job_template, Hash do
+            param :ansible_callback_enabled, :bool, :desc => N_('Enable the callback plugin for this template')
```
Comment (nofaralfasi) [inline app/controllers/foreman_ansible/api/v2/job_templates_controller_extensions.rb:12]: @adamruzicka same question here about the translations

File: app/views/job_templates/_job_template_callback_tab_content.html.erb (hunk 8 above)
Comment (adamruzicka) [reply app/views/job_templates/_job_template_callback_tab_content.html.erb:2]: Add translations as in "mark the strings for translation"? Yes

Usually marked strings are extracted by the maintainer outside of the changes which introduced the new strings so if you meant extracting strings or even translating them then that would be no

**Review by adamruzicka — APPROVED** [review]
I don't really have anything substantial to add here.

The migration could be made more efficient, but it doesn't really matter since in almost all cases we'll be updating a single record.

Codeclimate is not happy about this change, but it seems the consensus is to get rid of codeclimate completely so I'll go ahead and ignore it.

tl;dr: let's get this in

**Comment by adamruzicka** [comment]
Thank you @nofaralfasi !
